sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala

```diff
@@ -110,15 +116,43 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
   def inferField(typeSoFar: DataType, field: String): DataType = {
     if (field == null || field.isEmpty || field == options.nullValue) {
       typeSoFar
+    } else
+    if (options.inferDate) {
+      val typeElemInfer = typeSoFar match {
+        case NullType => tryParseInteger(field)
+        case IntegerType => tryParseInteger(field)
+        case LongType => tryParseLong(field)
+        case _: DecimalType => tryParseDecimal(field)
+        case DoubleType => tryParseDouble(field)
+        case DateType => tryParseDateTime(field)
+        case TimestampNTZType => tryParseDateTime(field)
+        case TimestampType => tryParseDateTime(field)
+        case BooleanType => tryParseBoolean(field)
+        case StringType => StringType
+        case other: DataType =>
+          throw QueryExecutionErrors.dataTypeUnexpectedError(other)
+      }
+      compatibleType(typeSoFar, typeElemInfer).getOrElse(StringType)
     } else {
       val typeElemInfer = typeSoFar match {
         case NullType => tryParseInteger(field)
         case IntegerType => tryParseInteger(field)
         case LongType => tryParseLong(field)
         case _: DecimalType => tryParseDecimal(field)
         case DoubleType => tryParseDouble(field)
-        case TimestampNTZType => tryParseTimestampNTZ(field)
-        case TimestampType => tryParseTimestamp(field)
+        case DateType => tryParseDateTime(field)
+        case TimestampNTZType =>
+          if (options.inferDate) {
```

Comment:
Hm, seems this logic cannot be reached, and duplicates the case-match above. Can we merge this if-else with above?

Reply:
My bad, I meant to remove the first match expression but missed it during code cleanup